hkmc2/shared/src/main/scala/hkmc2/codegen/Deforestation.scala

```diff
@@ -0,0 +1,1104 @@
+package hkmc2
+package codegen
+
+import semantics.*
+import semantics.Elaborator.State
+import syntax.{Literal, Tree}
+import utils.{TL, tl, SymbolSubst}
+import mlscript.utils.*, shorthands.*
+import scala.collection.mutable
+import scala.collection.mutable.LinkedHashMap
+
+type StratVar
+type StratVarId = Uid[StratVar]
+type ClsOrModSymbol = ClassLikeSymbol
+
+sealed abstract class Strat
+
+sealed abstract class ProdStrat extends Strat
+
+sealed abstract class ConsStrat extends Strat
+
+class StratVarState(val uid: StratVarId, val name: Str = ""):
+  lazy val asProdStrat = ProdVar(this)
+  lazy val asConsStrat = ConsVar(this)
+  
+  override def toString(): String = s"${if name.isEmpty() then "var" else name}@${uid}"
+
+object StratVarState:
+  
+  def freshVar(nme: String = "")(using vuid: Uid.Handler[StratVar]#State) =
+    val newId = vuid.nextUid
+    val s = StratVarState(newId, nme)
+    val p = s.asProdStrat
+    val c = s.asConsStrat
+    p -> c
+
+
+type CtorExpr = ResultId
```

Comment:
Why not use `ResultId` everywhere? `CtorExpr` gives the wrong impression that this is an actual expression node/AST.